### DISCUSSION TITLE
Build before debug

### DISCUSSION THREAD

**Discussion by grigorig** [xpack/vscode-xpack-extension-ts#32] [Ideas]
It would be very useful to be able to ensure that the executable is built and up to date before a debug session is started, e.g. with the Cortex Debug extension.
The "CMake Tools" extension allows this by referring to the executable via variable expansion, e.g. you specify `"executable": "${command:cmake.launchTargetPath}",` in the debug configuration. When the variable is expanded, this will implicitly launch a build before debugging is started. I don't see such a feature in the xPack extension, but I think it would be very useful.

**Comment by ilg-ul** [comment]
Hi @grigorig, you are right, it would be useful to trigger a build before debug. 

I don't know yet how to implement it, but I'll try to find out. 

Thank you for the suggestion.

**Reply by grigorig** [reply]
Actually, `preLaunchTask` in launch configuration works quite well. I know why I didn't try that before: CMake Tools doesn't expose any build tasks. The only advantage of the approach used by that extension is that it actually manages the currently selected build configuration. But I'm not convinced anymore having such a thing is a good idea.

Maybe this issue can be closed, not sure.

**Reply by ilg-ul** [reply]
> The only advantage of the approach used by that extension is that it actually manages the currently selected build configuration. But I'm not convinced anymore having such a thing is a good idea.

I'm afraid I'm not able to follow you. :-(

For my use case the CMake Tools extension did not work from the very beginning, and I did not further investigate.

**Reply by ilg-ul** [reply]
If you have a clear solution now on how to trigger a build before debug, please contribute it here as a more detailed message, so that those with less experience to be able to use it.

**Reply by grigorig** [reply]
OK, I guess I was a bit obtuse: the "CMake Tools" extension has a concept of a "currently selected build configuration". Similar to Eclipse CDT to some degree. With the implicit build-before-debug feature of that extension, you can share a single launch configuration for all build configurations quite easily.

With xPack, that's not really possible right now. You can only use the generic `preLaunchTask` feature to trigger a specific launch task. If you want to debug different build configurations, you need multiple launch configurations. That is good enough for me at the moment, though.

The thing is, I'm not so convinced anymore that having a "currently selected build configuration" is a good feature. I'd rather make sure that all configurations work all the time, and a "currently selected" configuration works against that. Things tend to break when they're not tested.

**Reply by ilg-ul** [reply]
> With the implicit build-before-debug feature of that extension, you can share a single launch configuration for all build configurations quite easily.

Ah, right. In Eclipse I never used it that way, I always configured launch configurations to build a specific build configuration.

> I'm not so convinced anymore that having a "currently selected build configuration" is a good feature

IntelliSense needs a way to know the current build configuration.

> I'd rather make sure that all configurations work all the time, and a "currently selected" configuration works against that. Things tend to break when they're not tested.

Sure. Here is how I use a [`test-all`](https://github.com/micro-os-plus/utils-lists-xpack/blob/616fb03b9835d1c1c17c3048c082f7ba0e005a1e/package.json#L72) action to test all build configurations on all platforms.

**Reply by ilg-ul** [reply]
> With xPack, that's not really possible right now.

My current use of xPack is to manage builds of multi-platform projects, mainly for running unit-tests. 

Debug is a slightly different use case; I plan to cover it soon, as I plan to cover parsing the build output for errors, to easily link to the wrong lines. Just that it'll take some time to finalise.

So any suggestions on how to improve things are highly appreciated.
